File: sync-models.ts
import { readFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import type { ExtensionAPI, ExtensionContext, ModelRegistry } from "@earendil-works/pi-coding-agent";
const HERE_DIR = resolve(dirname(fileURLToPath(import.meta.url)), ".");
const AGENT_DIR = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const SYNC_SCRIPT = resolve(HERE_DIR, "sync-models.py");
const MODELS_CONFIG = resolve(AGENT_DIR, "models.json");

interface ModelsConfig {
  providers?: Record<string, {
    baseUrl?: string;
    apiKey?: string;
    models?: Array<{
      id?: string;
      name?: string;
      thinkingLevelMap?: {
        off?: string | null;
      };
      compat?: {
        thinkingFormat?: string;
      };
      samplingParams?: {
        chat_template_kwargs?: {
          preserve_thinking?: boolean;
        };
      };
    }>;
  }>;
}

async function readModelsConfig(): Promise<ModelsConfig> {
  try {
    return JSON.parse(await readFile(MODELS_CONFIG, "utf8")) as ModelsConfig;
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") return {};
    throw error;
  }
}

interface ServerModel {
  id: string;
  /** Non-chat models are tagged (e.g. "text-to-image"); chat models omit this. */
  task?: string | null;
  quant?: string;
  loaded?: boolean;
  context_length?: number;
  max_context_length?: number;
  native_context_length?: number;
}

interface LoadedModelInfo {
  /** Provider key in models.json (needed to look the model up in the registry). */
  provider: string;
  /** Model id as registered in models.json (includes the quantization when present). */
  modelId: string;
  /** Model id as reported by the server, without quantization suffix. */
  serverId: string;
  name: string;
  contextLength: number | null;
  hasVision: boolean | null; // null = unknown
}

async function fetchLoadedModel(): Promise<LoadedModelInfo | null> {
  const config = await readModelsConfig();
  const providers = Object.entries(config.providers ?? {});
  // Prefer the provider pi is currently using, fall back to the first one.
  const [providerName, provider] = providers.find(([name]) => name === process.env.PI_PROVIDER) ?? providers[0] ?? [];
  if (!providerName || !provider) throw new Error("no providers in models.json");
  if (!provider?.baseUrl) throw new Error("provider has no baseUrl");

  let resp: Response;
  try {
    resp = await fetch(provider.baseUrl.replace(/\/$/, "") + "/models", {
      headers: { Authorization: `Bearer ${provider.apiKey ?? ""}` },
      signal: AbortSignal.timeout(5000),
    });
  } catch (error) {
    throw new Error(`cannot reach ${provider.baseUrl}: ${error instanceof Error ? error.message : String(error)}`);
  }
  if (!resp.ok) throw new Error(`server returned HTTP ${resp.status}`);

  const payload = (await resp.json()) as { data?: ServerModel[] };
  const chatTasks = new Set([undefined, null, "text-generation", "chat", "completion"]);
  const loaded = (payload.data ?? []).find((model) => model.loaded && chatTasks.has(model.task));
  if (!loaded) return null; // nothing in memory — legitimately empty

  const id = loaded.quant ? `${loaded.id}:${loaded.quant}` : loaded.id;
  const entry =
    (provider.models ?? []).find((model) => model.id === id) ??
    (provider.models ?? []).find((model) => model.id === loaded.id);

  let contextLength: number | null = null;
  // On unsloth-studio, `context_length` is the context actually allocated in memory;
  // `max_context_length` mirrors the active --reasoning-budget (set via /reasoning-budget apply),
  // and `native_context_length` is the GGUF-trained max. None of those are the loaded context.
  const fields = ["context_length", "max_context_length", "native_context_length"] as const;
  for (const field of fields) {
    const value = loaded[field];
    if (typeof value === "number" && value > 0) {
      contextLength = value;
      break;
    }
  }

  // Best effort: Studio's /api/models/list reports is_vision per model.
  let hasVision: boolean | null = null;
  try {
    const base = new URL(provider.baseUrl);
    const listResp = await fetch(`${base.protocol}//${base.host}/api/models/list`, {
      headers: { Authorization: `Bearer ${provider.apiKey ?? ""}` },
      signal: AbortSignal.timeout(5000),
    });
    if (listResp.ok) {
      const list = (await listResp.json()) as { models?: Array<{ id?: string; is_vision?: boolean }> };
      hasVision = (list.models ?? []).find((model) => model.id === loaded.id)?.is_vision ?? null;
    }
  } catch {
    // ignore — vision info is optional
  }

  return {
    provider: providerName,
    modelId: entry?.id ?? id,
    serverId: loaded.id,
    name: entry?.name ?? entry?.id ?? id,
    contextLength,
    hasVision,
  };
}

async function qwenThinkingIsValid(): Promise<boolean> {
  const config = await readModelsConfig();
  const qwenModels = Object.values(config.providers ?? {}).flatMap((provider) =>
    (provider.models ?? []).filter((model) => model.id?.toLowerCase().includes("qwen3.8"))
  );
  return qwenModels.length === 0 || qwenModels.every((model) =>
    model.thinkingLevelMap?.off === "none"
    && model.compat?.thinkingFormat === "qwen"
    && model.samplingParams?.chat_template_kwargs?.preserve_thinking === true
  );
}

type PiModel = NonNullable<ReturnType<ModelRegistry["find"]>>;

function normalize(value: string): string {
  return value.toLowerCase();
}

/** Candidate ids for the loaded model: quantized id first, then the bare server id. */
function candidateIds(loaded: LoadedModelInfo): string[] {
  const candidates: string[] = [];
  for (const id of [loaded.modelId, loaded.serverId]) {
    if (!candidates.some((seen) => normalize(seen) === normalize(id))) candidates.push(id);
  }
  return candidates;
}

function isCurrentModel(model: PiModel | undefined, loaded: LoadedModelInfo): boolean {
  if (!model || model.provider !== loaded.provider) return false;
  const current = normalize(model.id);
  return candidateIds(loaded).some((id) => normalize(id) === current);
}

function findRegistryModel(registry: ModelRegistry, loaded: LoadedModelInfo): PiModel | undefined {
  for (const id of candidateIds(loaded)) {
    const model = registry.find(loaded.provider, id);
    if (model) return model;
  }
  // Fall back to a case-insensitive scan of the available catalogue.
  const candidates = candidateIds(loaded).map(normalize);
  return registry
    .getAvailable()
    .find((model) => model.provider === loaded.provider && candidates.includes(normalize(model.id)));
}

type LoadedReport = { loaded: LoadedModelInfo; label: string };

async function describeLoadedModel(ctx: ExtensionContext): Promise<LoadedReport | null> {
  let loaded: LoadedModelInfo | null;
  try {
    loaded = await fetchLoadedModel();
  } catch (error) {
    ctx.ui.notify(`Loaded model: unavailable (${error instanceof Error ? error.message : String(error)})`, "error");
    return null;
  }
  if (!loaded) {
    ctx.ui.notify("No model loaded on server", "warning");
    return null;
  }
  const details = [
    loaded.contextLength ? `context ${loaded.contextLength}` : null,
    loaded.hasVision === null ? null : loaded.hasVision ? "vision" : "text-only",
  ].filter(Boolean).join(", ");
  return { loaded, label: `${loaded.name}${details ? ` (${details})` : ""}` };
}

async function showLoadedModel(ctx: ExtensionContext): Promise<void> {
  const report = await describeLoadedModel(ctx);
  if (report) ctx.ui.notify(`Loaded model: ${report.label}`, "info");
}

/** Switch the session to whichever model is currently resident in server memory. */
async function useLoadedModel(pi: ExtensionAPI, ctx: ExtensionContext): Promise<void> {
  const report = await describeLoadedModel(ctx);
  if (!report) return;
  const { loaded, label } = report;

  if (isCurrentModel(ctx.model, loaded)) {
    ctx.ui.notify(`Loaded model: ${label} (already active)`, "info");
    return;
  }

  const model = findRegistryModel(ctx.modelRegistry, loaded);
  if (!model) {
    ctx.ui.notify(`Loaded model: ${label} — no matching entry in models.json. Run /sync-models to add it`, "warning");
    return;
  }

  if (ctx.scopedModels.length > 0 && !ctx.scopedModels.some((scoped) => scoped.model.id === model.id)) {
    ctx.ui.notify(`Loaded model: ${label} is outside this session's model scope`, "warning");
    return;
  }

  if (await pi.setModel(model)) {
    ctx.ui.notify(`Switched to loaded model: ${label}`, "info");
  } else {
    ctx.ui.notify(`Cannot switch to ${label}: no credentials for provider "${model.provider}"`, "error");
  }
}

export default function (pi: ExtensionAPI) {
  let syncing = false;

  pi.on("session_start", (event, ctx) => {
    if (event.reason !== "startup") return;
    // Fire-and-forget so startup is not blocked by the server fetch.
    void useLoadedModel(pi, ctx);
  });

  pi.registerCommand("model-loaded", {
    description: 'Show which model is loaded in server memory ("use" switches to it)',
    getArgumentCompletions: (prefix) => {
      const option = { value: "use", label: "use", description: "Switch the session to the loaded model" };
      return option.value.startsWith(prefix.trim().toLowerCase()) ? [option] : null;
    },
    handler: async (args, ctx) => {
      const argument = args.trim().toLowerCase();
      if (!argument) await showLoadedModel(ctx);
      else if (argument === "use") await useLoadedModel(pi, ctx);
      else ctx.ui.notify("Usage: /model-loaded [use]", "error");
    },
  });

  pi.registerCommand("sync-models", {
    description: "Pull the latest server models (one entry per downloaded GGUF quantization) into models.json and reload Pi",
    getArgumentCompletions: (prefix) => {
      const option = { value: "all", label: "all", description: "Include embedding models" };
      return option.value.startsWith(prefix.trim().toLowerCase()) ? [option] : null;
    },
    handler: async (args, ctx) => {
      const argument = args.trim().toLowerCase();
      if (argument && argument !== "all") {
        ctx.ui.notify("Usage: /sync-models [all]", "error");
        return;
      }

      if (syncing) {
        ctx.ui.notify("Model synchronization is already running", "warning");
        return;
      }

      syncing = true;
      const scriptArgs = [SYNC_SCRIPT, "--config", MODELS_CONFIG];
      if (argument === "all") scriptArgs.push("--include-all");

      try {
        const config = await readModelsConfig();
        const providers = Object.entries(config.providers ?? {});
        if (providers.length > 1) {
          throw new Error("models.json must contain at most one provider");
        }

        const promptRequired = async (title: string, placeholder: string): Promise<string> => {
          const value = (await ctx.ui.input(title, placeholder))?.trim();
          if (!value) throw new Error("Setup cancelled");
          return value;
        };

        if (providers.length === 0) {
          const providerName = await promptRequired("Provider name", "thread");
          scriptArgs.push("--provider-name", providerName);
        }

        const provider = providers[0]?.[1];
        if (!provider?.baseUrl) {
          const ip = await promptRequired("Server IP or hostname", "192.168.1.10");
          const port = await promptRequired("Server port", "8888");
          const portNumber = Number(port);
          if (!Number.isInteger(portNumber) || portNumber < 1 || portNumber > 65535) {
            throw new Error("Server port must be between 1 and 65535");
          }
          scriptArgs.push("--ip", ip, "--port", String(portNumber));
        }

        if (!provider?.apiKey) {
          const apiKey = await promptRequired("API key", "sk-...");
          scriptArgs.push("--api-key", apiKey);
        }
      } catch (error) {
        syncing = false;
        const message = error instanceof Error ? error.message : String(error);
        ctx.ui.notify(message, message === "Setup cancelled" ? "warning" : "error");
        return;
      }

      ctx.ui.notify("Synchronizing models…", "info");

      let result;
      try {
        result = await pi.exec("python3", scriptArgs, {
          cwd: AGENT_DIR,
          timeout: 30_000,
        });
      } catch (error) {
        syncing = false;
        const message = error instanceof Error ? error.message : String(error);
        ctx.ui.notify(`Model synchronization failed: ${message}`, "error");
        return;
      }

      if (result.code !== 0) {
        syncing = false;
        const error = result.stderr.trim() || result.stdout.trim() || `sync exited with code ${result.code}`;
        ctx.ui.notify(`Model synchronization failed: ${error}`, "error");
        return;
      }

      if (!(await qwenThinkingIsValid())) {
        syncing = false;
        ctx.ui.notify("Model synchronization failed: Qwen3.8 thinking configuration is incomplete", "error");
        return;
      }

      syncing = false;
      const [header, ...modelIds] = result.stdout.trim().split("\n").map((line) => line.trim());
      const question = `Reload Pi now to apply?`;
      const body = [header, "", ...modelIds].join("\n");
      if (await ctx.ui.confirm("Models synchronized", `${body}\n\n${question}`)) await ctx.reload();
      else ctx.ui.notify("Reload cancelled — new models will apply on next restart", "info");
      return;
    },
  });
}

File: sync-models.py
#!/usr/bin/env python3
"""Synchronize Pi's model list from an OpenAI-compatible server.

Usage:
    .pi/agent/extensions/sync-models.py
    .pi/agent/extensions/sync-models.py --include-all
    .pi/agent/extensions/sync-models.py --api-key KEY --ip 192.168.1.10 --port 8888
    .pi/agent/extensions/sync-models.py --config /path/to/models.json

The existing provider configuration is preserved. Existing model entries keep
their Pi-specific metadata; new entries receive sensible defaults based on
the server's /v1/models response. Missing configuration is created
interactively.

GGUF models with more than one downloaded quantization get one entry per
quant, addressed as "<model id>:<quant>" (e.g.
"unsloth/Qwen3.8-27B-GGUF:UD-Q8_K_XL"). With Studio's OpenAI auto-switch
enabled, requesting such an id loads/switches to that quantization.
Models with a single quant keep their plain id.
"""

from __future__ import annotations

import argparse
import getpass
import json
import os
import re
import sys
import tempfile
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.parse import quote, urlsplit, urlunsplit
from urllib.request import Request, urlopen


DEFAULT_CONFIG = Path(__file__).resolve().parent.parent / "models.json"
DEFAULT_CONTEXT = 128_000
DEFAULT_MAX_TOKENS = 32_000
DEFAULT_PORT = 8888


def fetch_models(base_url: str, api_key: str) -> list[dict]:
    url = base_url.rstrip("/") + "/models"
    request = Request(url, headers={"Authorization": f"Bearer {api_key}"})
    try:
        with urlopen(request, timeout=15) as response:
            payload = json.load(response)
    except HTTPError as exc:
        raise RuntimeError(f"server returned HTTP {exc.code} for {url}") from exc
    except URLError as exc:
        raise RuntimeError(f"could not connect to {url}: {exc.reason}") from exc
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"server returned invalid JSON from {url}") from exc

    models = payload.get("data")
    if not isinstance(models, list):
        raise RuntimeError("model response does not contain a 'data' list")
    return [model for model in models if isinstance(model, dict) and model.get("id")]


# Studio tags non-LLM models with a `task` field (e.g. Z-Image-Turbo-GGUF is
# "text-to-image"); chat models have no task or a plain text-generation one.
CHAT_TASKS = {None, "text-generation", "chat", "completion"}


def is_non_chat(model: dict) -> bool:
    return model.get("task") not in CHAT_TASKS


def is_embedding(model: dict) -> bool:
    haystack = " ".join(
        str(model.get(field, "")) for field in ("id", "display_name", "owned_by")
    ).lower()
    return "embedding" in haystack or "embed" in haystack


def context_window(model: dict) -> int:
    # Prefer the native limit, then the advertised maximum/current limit.
    for field in ("native_context_length", "max_context_length", "context_length"):
        value = model.get(field)
        if isinstance(value, int) and value > 0:
            return value
    return DEFAULT_CONTEXT


def fetch_quant_variants(base_url: str, api_key: str, model_id: str) -> list[str]:
    """Return the downloaded GGUF quantizations for a model repo (best effort).

    Uses Studio's /api/models/gguf-variants endpoint with
    prefer_local_cache=true: the network-resolved listing reports multi-part GGUF
    repos (e.g. Qwen3.8-Flash-Next) as never downloaded even with every shard on
    disk, and this flag answers from the local cache instead. The endpoint lives
    on the server root rather than under the /v1 base path. Returns [] when it is
    unavailable or the repo has no variant list.
    """
    parts = urlsplit(base_url.rstrip("/"))
    root = urlunsplit((parts.scheme, parts.netloc, "", "", ""))
    url = (
        f"{root}/api/models/gguf-variants"
        f"?repo_id={quote(model_id, safe='')}&prefer_local_cache=true"
    )
    request = Request(url, headers={"Authorization": f"Bearer {api_key}"})
    try:
        with urlopen(request, timeout=15) as response:
            payload = json.load(response)
    except (HTTPError, URLError, json.JSONDecodeError):
        return []
    variants = payload.get("variants") if isinstance(payload, dict) else None
    if not isinstance(variants, list):
        return []
    quants: list[str] = []
    for variant in variants:
        if not isinstance(variant, dict) or not variant.get("downloaded"):
            continue
        quant = variant.get("quant")
        if isinstance(quant, str) and quant not in quants:
            quants.append(quant)
    return quants


def derive_name(model_id: str) -> str:
    name = model_id.split("/")[-1]
    if name.upper().endswith("-GGUF"):
        name = name[: -len("-GGUF")]
    return name.replace("-", " ")


def make_entry(
    model_id: str,
    old_by_id: dict[str, dict],
    server_model: dict | None = None,
    base_model_id: str | None = None,
    quant: str | None = None,
) -> dict:
    # Inherit metadata from an entry with the exact id first, then from the
    # base model entry (so plain-id config carries over to quant-suffixed ids).
    old = old_by_id.get(model_id)
    if old is None and base_model_id:
        old = old_by_id.get(base_model_id)
    old = dict(old or {})
    server_model = server_model or {}
    entry = dict(old)
    entry["id"] = model_id
    entry.setdefault("reasoning", False)
    entry.setdefault("input", ["text"])
    entry.setdefault("contextWindow", context_window(server_model))
    entry.setdefault("maxTokens", DEFAULT_MAX_TOKENS)
    entry.setdefault(
        "cost",
        {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0},
    )
    if "qwen3.8" in model_id.lower():
        entry["reasoning"] = True
        entry["thinkingLevelMap"] = {
            "off": "none",
            "minimal": "minimal",
            "low": "low",
            "medium": "medium",
            "high": "high",
            "xhigh": "xhigh",
        }
        compat = dict(entry.get("compat") or {})
        compat["thinkingFormat"] = "qwen"
        entry["compat"] = compat
        sampling_params = dict(entry.get("samplingParams") or {})
        chat_template_kwargs = dict(
            sampling_params.get("chat_template_kwargs") or {}
        )
        chat_template_kwargs["preserve_thinking"] = True
        sampling_params["chat_template_kwargs"] = chat_template_kwargs
        entry["samplingParams"] = sampling_params
    if quant:
        base_name = old.get("name") or derive_name(base_model_id or model_id)
        # Drop any stale "(quant)" suffix left over from a previous sync.
        base_name = re.sub(r"\s*\([^)]*\)$", "", str(base_name)).strip()
        entry["name"] = f"{base_name} ({quant})"
    return entry


def write_json_atomically(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temporary_name = tempfile.mkstemp(prefix=f".{path.name}.", dir=path.parent)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as output:
            json.dump(data, output, indent=2)
            output.write("\n")
        os.replace(temporary_name, path)
    except Exception:
        try:
            os.unlink(temporary_name)
        except FileNotFoundError:
            pass
        raise


def prompt_value(label: str, default: str | None = None, secret: bool = False) -> str:
    suffix = f" [{default}]" if default else ""
    try:
        value = getpass.getpass(f"{label}{suffix}: ") if secret else input(f"{label}{suffix}: ")
    except (EOFError, KeyboardInterrupt) as exc:
        raise RuntimeError(f"{label} is required") from exc
    value = value.strip() or (default or "")
    if not value:
        raise RuntimeError(f"{label} is required")
    return value


def load_config(path: Path) -> dict:
    if not path.exists():
        return {"providers": {}}
    with path.open(encoding="utf-8") as source:
        config = json.load(source)
    if not isinstance(config, dict):
        raise RuntimeError("models config must contain a JSON object")
    return config


def configure_provider(config: dict, args: argparse.Namespace) -> tuple[str, dict, str, str]:
    providers = config.setdefault("providers", {})
    if not isinstance(providers, dict):
        raise RuntimeError("models config 'providers' must be an object")

    if len(providers) > 1:
        raise RuntimeError("expected at most one provider in the config")

    if providers:
        provider_name, provider = next(iter(providers.items()))
        if not isinstance(provider, dict):
            raise RuntimeError(f"provider '{provider_name}' must be an object")
    else:
        provider_name = args.provider_name or prompt_value("Provider name")
        provider = {
            "api": "openai-completions",
            "compat": {
                "supportsDeveloperRole": False,
                "supportsReasoningEffort": True,
            },
            "models": [],
        }
        providers[provider_name] = provider

    base_url = provider.get("baseUrl")
    parsed = urlsplit(base_url) if isinstance(base_url, str) and base_url else None
    hostname = args.ip or (parsed.hostname if parsed else None)
    port = args.port or (parsed.port if parsed else None)

    if not hostname:
        hostname = prompt_value("Server IP or hostname")
    if port is None:
        port_text = prompt_value("Server port", str(DEFAULT_PORT))
        try:
            port = int(port_text)
        except ValueError as exc:
            raise RuntimeError("Server port must be an integer") from exc
    if not 1 <= port <= 65535:
        raise RuntimeError("Server port must be between 1 and 65535")

    scheme = parsed.scheme if parsed else "http"
    path = parsed.path if parsed and parsed.path else "/v1"
    base_url = urlunsplit((scheme, f"{hostname}:{port}", path, "", ""))

    api_key = args.api_key or provider.get("apiKey")
    if not isinstance(api_key, str) or not api_key:
        api_key = prompt_value("API key", secret=True)

    provider["baseUrl"] = base_url
    provider["apiKey"] = api_key
    return provider_name, provider, base_url, api_key


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--config", type=Path, default=DEFAULT_CONFIG)
    parser.add_argument("--api-key", help="override the API key from models.json")
    parser.add_argument("--ip", help="override the Unsloth server IP/hostname")
    parser.add_argument("--port", type=int, help="override the Unsloth server port")
    parser.add_argument("--provider-name", help="provider name when creating models.json")
    parser.add_argument(
        "--include-all",
        action="store_true",
        help="include embedding models as well as chat/completion models",
    )
    args = parser.parse_args()

    try:
        config = load_config(args.config)
        provider_name, provider, base_url, api_key = configure_provider(config, args)
        write_json_atomically(args.config, config)

        server_models = fetch_models(base_url, api_key)
        server_models = [model for model in server_models if not is_non_chat(model)]
        if not args.include_all:
            server_models = [model for model in server_models if not is_embedding(model)]
        if not server_models:
            raise RuntimeError("server returned no usable models; refusing to overwrite config")

        old_models = provider.get("models", [])
        old_by_id = {
            model["id"]: model
            for model in old_models
            if isinstance(model, dict) and model.get("id")
        }

        entries: list[dict] = []
        for model in server_models:
            model_id = model["id"]
            quants: list[str] = []
            if isinstance(model_id, str) and model_id.upper().endswith("-GGUF"):
                quants = fetch_quant_variants(base_url, api_key, model_id)
                loaded = model.get("quant")
                # Keep the currently loaded quant first so the default
                # position matches what the server has in memory.
                if isinstance(loaded, str) and loaded in quants:
                    quants.remove(loaded)
                    quants.insert(0, loaded)
            if len(quants) > 1:
                for quant in quants:
                    entries.append(
                        make_entry(
                            f"{model_id}:{quant}",
                            old_by_id,
                            model,
                            base_model_id=model_id,
                            quant=quant,
                        )
                    )
            else:
                entries.append(make_entry(model_id, old_by_id, model))
        provider["models"] = entries
        write_json_atomically(args.config, config)

        print(f"Synchronized {len(provider['models'])} models in {args.config}")
        for model in provider["models"]:
            print(f"  {model['id']}")
        return 0
    except (OSError, ValueError, RuntimeError) as exc:
        print(f"sync-models: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
